File: LeetCode/637AverageOfLevelsInBinaryTree/AverageOfLevelsInBinaryTree.py
class Solution:
    def averageOfLevels(self, root: Optional[TreeNode]) -> List[float]:
        # BFS -> Time:O(n), Space:O(2^h)
        q = []
        q.append(root)
        ans = []
        while q:
            size = len(q)
            level_sum = 0
            for i in range(size):
                node = q.pop(0)
                level_sum += node.val
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
            ans.append(level_sum/size)
        return ans

File: LeetCode/637AverageOfLevelsInBinaryTree/AverageOfLevelsInBinaryTree.cpp
class Solution {
public:
    vector<double> averageOfLevels(TreeNode* root) {
        // BFS -> Time:O(n), Space:O(2^h)
        queue<TreeNode*> q;
        q.push(root);
        int size;
        double levelSum;
        TreeNode *node;
        vector<double> ans;
        while (!q.empty()) {
            size = q.size();
            levelSum = 0;
            for (int i = 0; i < size; i++) {
                node = q.front();
                q.pop();
                levelSum += node->val;
                if (node->left) {
                    q.push(node->left);
                }
                if (node->right) {
                    q.push(node->right);
                }
            }
            ans.push_back(levelSum/size);
        }
        return ans;
    }
};
